tsfc/driver.py

```diff
@@ -252,10 +255,10 @@ def name_multiindex(multiindex, name):
     body = generate_coffee(impero_c, index_names, parameters["precision"],
                            parameters["scalar_type"], expressions, split_argument_indices)
 
-    return builder.construct_kernel(kernel_name, body)
+    return builder.construct_kernel(kernel_name, body, quad_rule)
```

Comment:
This is kind of bugged: `quad_rule` comes from a loop over quadratures, but here outside the loop it only references the value in the last iteration. It works as long as there is only one quadrature in an integral.

Reply:
This is very true! It We only need quad_rule in order to figure out which points to tabulate at for runtime tabulation. A better way to do this would be to somehow attach this information to the tabulations themselves. I am not sure how to do this though...

Reply:
So what is the kernel interface right now? You make a quad rule so that TSFC knows what size the tabulations are (and that's it?). What part of themis uses the quad rule that comes back in `construct_kernel`?

Reply:
I think TSFC already knows the size of the tabulations (they are just GEM nodes?). Themis uses the quad_rule to determine which points to tabulate at (using the shift axis and restriction), since the tabulations don't actually know this....

Reply:
Well, what happens is that the quad rule comes in (with some points). This point set is provided to the element for tabulation, which constructs the shape of the tabulation matrix like so:

```python
        shape = tuple(index.extent for index in ps.indices) + self.index_shape + self.value_shape
```

where `ps.indices` are the indices used to index the quadrature rule's point set.

Reply:
Is there a way to associate these points with the tabulation (specifically the GEM node) within TSFC/FInAT?

Reply:
Right now, FInAT tabulates like this:

```python
        for derivative in range(order + 1):
            for alpha in mis(dimension, derivative):
                name = str.format("rt_{}_{}_{}_{}_{}_{}",
                                  self.variant,
                                  self.degree,
                                  ''.join(map(str, alpha)),
                                  self.shift_axes,
                                  'c' if self.continuous else 'd',
                                  {None: "",
                                   '+': "p",
                                   '-': "m"}[self.restriction])
                result[alpha] = gem.partial_indexed(gem.Variable(name, shape), ps.indices)
```

You would, I suppose, need to introduce a new node in gem. Say `gem.Tabulation`? that behaves like a variable, but carries that information. Arguably, this might things cleaner because the encoding of the tabulation information now doesn't live in the string name (I don't know if you use that at all).